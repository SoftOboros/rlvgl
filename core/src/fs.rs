//! Asset loading interfaces for filesystem-backed content.
//!
//! This module provides traits used by the optional `fs` feature to source
//! assets such as fonts or images from an underlying filesystem.

use alloc::boxed::Box;

/// Errors that can occur during filesystem operations.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FsError {
    /// Underlying device reported an error.
    Device,
    /// Provided path was invalid.
    InvalidPath,
    /// File or directory was not found.
    NoSuchFile,
}

/// Block device abstraction used by the filesystem layer.
///
/// Implementors provide sector-based access to a storage medium.
pub trait BlockDevice {
    /// Read blocks starting at `lba` into `buf`.
    fn read_blocks(&mut self, lba: u64, buf: &mut [u8]) -> Result<(), FsError>;

    /// Write blocks starting at `lba` from `buf`.
    ///
    /// Implementations may leave this unimplemented if the device is read-only.
    fn write_blocks(&mut self, lba: u64, buf: &[u8]) -> Result<(), FsError>;

    /// Return the logical block size in bytes.
    fn block_size(&self) -> usize;

    /// Return the total number of addressable blocks.
    fn num_blocks(&self) -> u64;

    /// Flush any buffered data to the underlying device.
    fn flush(&mut self) -> Result<(), FsError>;
}

/// Error type returned by asset operations.
#[derive(Debug, Clone)]
pub enum AssetError {
    /// Underlying filesystem error.
    Fs(FsError),
}

/// Reader trait for streaming asset data.
pub trait AssetRead {
    /// Read data into `out`, returning the number of bytes read.
    fn read(&mut self, out: &mut [u8]) -> Result<usize, AssetError>;

    /// Total length of the asset in bytes.
    fn len(&self) -> usize;

    /// Return `true` if the asset has a length of zero bytes.
    fn is_empty(&self) -> bool;

    /// Seek to an absolute byte position within the asset.
    fn seek(&mut self, pos: u64) -> Result<u64, AssetError>;
}

/// Source of assets such as fonts or images.
pub trait AssetSource {
    /// Open an asset by logical path, e.g., `"fonts/regular.bin"`.
    fn open<'a>(&'a self, path: &str) -> Result<Box<dyn AssetRead + 'a>, AssetError>;

    /// Determine whether an asset at `path` exists.
    fn exists(&self, path: &str) -> bool;

    /// List the contents of `dir`, returning an iterator over asset entries.
    fn list(&self, dir: &str) -> Result<AssetIter, AssetError>;
}

/// Iterator over asset entries returned by [`AssetSource::list`].
pub struct AssetIter;

impl Iterator for AssetIter {
    type Item = (); // placeholder until fleshed out

    fn next(&mut self) -> Option<Self::Item> {
        None
    }
}

/// Manager that provides convenient typed loading helpers.
pub struct AssetManager<S: AssetSource> {
    source: S,
}

impl<S: AssetSource> AssetManager<S> {
    /// Create a new [`AssetManager`] from an [`AssetSource`].
    pub fn new(source: S) -> Self {
        Self { source }
    }

    /// Open a raw asset stream at `path`.
    pub fn open(&self, path: &str) -> Result<Box<dyn AssetRead + '_>, AssetError> {
        self.source.open(path)
    }
}
