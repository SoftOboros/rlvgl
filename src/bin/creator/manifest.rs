//! Manifest structures for rlvgl-creator.
//!
//! Defines the `Manifest` schema used by CLI subcommands along with supporting
//! types for assets, groups, packages, and targets.

use std::collections::BTreeMap;

use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

/// Manifest describing assets and related metadata.
#[derive(Serialize, Deserialize, JsonSchema)]
pub(crate) struct Manifest {
    #[serde(default = "default_manifest_version")]
    pub(crate) version: u8,
    #[serde(default)]
    pub(crate) packages: BTreeMap<String, Package>,
    #[serde(default)]
    pub(crate) groups: BTreeMap<String, Group>,
    #[serde(default)]
    pub(crate) features: BTreeMap<String, Vec<String>>,
    #[serde(default)]
    pub(crate) expose: BTreeMap<String, String>,
    #[serde(default)]
    pub(crate) targets: Vec<Target>,
    #[serde(default)]
    pub(crate) assets: Vec<Asset>,
}

fn default_manifest_version() -> u8 {
    1
}

impl Default for Manifest {
    fn default() -> Self {
        Self {
            version: default_manifest_version(),
            packages: BTreeMap::new(),
            groups: BTreeMap::new(),
            features: BTreeMap::new(),
            expose: BTreeMap::new(),
            targets: Vec::new(),
            assets: Vec::new(),
        }
    }
}

/// Asset entry with path, hash, and generated constant name.
#[derive(Serialize, Deserialize, JsonSchema)]
pub(crate) struct Asset {
    /// Generated constant or feature name.
    #[serde(default)]
    pub(crate) name: String,
    /// Relative path to the asset.
    pub(crate) path: String,
    /// Blake3 hash of the asset contents.
    pub(crate) hash: String,
    /// SPDX license identifier for this asset.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) license: Option<String>,
}

/// Group of assets sharing an optional license.
#[derive(Default, Serialize, Deserialize, JsonSchema)]
pub(crate) struct Group {
    /// Assets belonging to this group.
    #[serde(default)]
    pub(crate) assets: Vec<String>,
    /// SPDX license identifier applying to the group.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub(crate) license: Option<String>,
}

/// Target output configuration.
#[derive(Serialize, Deserialize, JsonSchema)]
pub(crate) struct Target {
    /// Target name.
    pub(crate) name: String,
    /// Directory where assets will be vendored.
    pub(crate) vendor_dir: String,
}

/// Feature entry emitted by the `sync` command.
#[derive(Serialize)]
pub(crate) struct FeatureEntry {
    /// Feature name.
    pub(crate) name: String,
    /// Dependent features.
    pub(crate) deps: Vec<String>,
}

/// Package grouping helper.
#[derive(Default, Serialize, Deserialize, JsonSchema)]
pub(crate) struct Package {
    /// Assets in this package.
    #[serde(default)]
    pub(crate) assets: Vec<String>,
}
