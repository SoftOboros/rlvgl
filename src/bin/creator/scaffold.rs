//! Scaffold command for rlvgl-creator.
//!
//! Generates a dual-mode assets crate using embedded Tera templates.

use std::collections::BTreeMap;
use std::fs;
use std::path::Path;

use anyhow::{Result, bail};
use tera::{Context, Tera};

use crate::manifest::Manifest;
use serde_yaml;

/// Generate a new assets crate populated from the manifest.
pub(crate) fn run(path: &Path, manifest_path: &Path) -> Result<()> {
    if path.exists() {
        bail!("`{}` already exists", path.display());
    }

    let crate_name = path
        .file_name()
        .ok_or_else(|| anyhow::anyhow!("invalid crate path"))?
        .to_string_lossy()
        .into_owned();

    fs::create_dir_all(path.join("src"))?;

    let manifest: Manifest = if manifest_path.exists() {
        let contents = fs::read_to_string(manifest_path)?;
        serde_yaml::from_str(&contents)?
    } else {
        Manifest::default()
    };

    let mut groups: Vec<String> = manifest.groups.keys().map(|g| g.to_lowercase()).collect();
    groups.sort();

    // Map each asset path to the groups that reference it so constants can be feature gated.
    let mut per_asset_groups: BTreeMap<String, Vec<String>> = BTreeMap::new();
    for (group, info) in &manifest.groups {
        for asset in &info.assets {
            per_asset_groups
                .entry(asset.clone())
                .or_default()
                .push(group.to_lowercase());
        }
    }

    #[derive(serde::Serialize)]
    struct EmbedAsset {
        name: String,
        path: String,
        features: Vec<String>,
    }

    let mut embed_assets = Vec::new();
    for asset in &manifest.assets {
        embed_assets.push(EmbedAsset {
            name: asset.name.clone(),
            path: asset.path.clone(),
            features: per_asset_groups
                .get(&asset.path)
                .cloned()
                .unwrap_or_default(),
        });
    }

    let mut tera = Tera::default();
    const CARGO_TOML: &str = r#"[package]
name = "{{ name }}"
version = "0.1.0"
edition = "2024"
description = "Assets crate generated by rlvgl-creator"

[features]
default = []
embed = []
vendor = []
{% for g in groups %}{{ g }} = []
{% endfor %}

[dependencies]
phf = { version = "0.11", features = ["macros"], default-features = false }
"#;
    const LIB_RS: &str = r#"#![no_std]
#![deny(missing_docs)]
//! Assets crate generated by rlvgl-creator.

#[cfg(feature = "vendor")]
extern crate std;

#[cfg(feature = "embed")]
/// Embedded asset bytes.
pub mod embed {
    use phf::{phf_map, Map};

    {% for a in assets %}
    {% if a.features | length > 0 %}
    #[cfg(any({% for f in a.features %}feature = "{{ f }}"{% if not loop.last %}, {% endif %}{% endfor %}))]
    {% endif %}
    pub const {{ a.name }}: &[u8] = include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/../{{ a.path }}"));
    {% endfor %}

    static INDEX: Map<&'static str, &'static [u8]> = phf_map! {
    {% for a in assets %}
        "{{ a.path }}" => {{ a.name }},
    {% endfor %}
    };

    /// Fetch an embedded asset by its relative path.
    pub fn get(path: &str) -> Option<&'static [u8]> {
        INDEX.get(path).copied()
    }
}

#[cfg(feature = "vendor")]
/// Vendor build helpers.
pub mod vendor {
    use std::path::Path;

    /// Copy all assets into `out_dir`.
    pub fn copy_all(out_dir: &Path) -> std::io::Result<()> {
        vendor_api::copy_all(out_dir, ASSETS)
    }

    /// Generate an `rlvgl_assets.rs` module under `out_dir`.
    pub fn generate_rust_module(out_dir: &Path) -> std::io::Result<()> {
        vendor_api::generate_rust_module(out_dir, ASSETS)
    }

    const ASSETS: &[(&str, &str)] = &[
    {% for a in assets %}
        ("{{ a.name }}", "{{ a.path }}"),
    {% endfor %}
    ];

    mod vendor_api {
        use std::{fs, path::Path, string::String, format};

        pub fn copy_all(out_dir: &Path, assets: &[(&str, &str)]) -> std::io::Result<()> {
            let root = Path::new(env!("CARGO_MANIFEST_DIR")).join("../");
            for (_, path) in assets {
                let src = root.join(path);
                let dest = out_dir.join(path);
                if let Some(parent) = dest.parent() {
                    fs::create_dir_all(parent)?;
                }
                fs::copy(&src, &dest)?;
            }
            Ok(())
        }

        pub fn generate_rust_module(out_dir: &Path, assets: &[(&str, &str)]) -> std::io::Result<()> {
            let mut module = String::from("//! Auto-generated asset constants\n\n");
            for (name, path) in assets {
                module.push_str(&format!("pub const {}: &[u8] = include_bytes!(\"{}\");\n", name, path));
            }
            fs::write(out_dir.join("rlvgl_assets.rs"), module)?;
            Ok(())
        }
    }
}
"#;
    const README_MD: &str = r#"# {{ name }}

This crate was generated by `rlvgl-creator`.

## Embed

Enable the `embed` feature and reference asset constants directly:

```rust
use {{ name }}::embed::SOME_ASSET;
```

## Vendor

Enable the `vendor` feature to have the included `build.rs` copy asset files
and generate an `rlvgl_assets.rs` module at compile time. In your crate,
include the generated module:

```rust
include!(concat!(env!("OUT_DIR"), "/rlvgl_assets.rs"));
```

For custom pipelines, you may call the helpers from your own `build.rs`:

```rust
fn main() {
    let out = std::path::Path::new(&std::env::var("OUT_DIR").unwrap());
    {{ name }}::vendor::copy_all(out).unwrap();
    {{ name }}::vendor::generate_rust_module(out).unwrap();
}
```
"#;
    const BUILD_RS: &str = r#"// Build script for the assets crate.
//
// When built with the `vendor` feature this script copies asset files into the
// build output directory and generates an `rlvgl_assets.rs` module with
// `include_bytes!` declarations.

fn main() -> std::io::Result<()> {
    #[cfg(feature = "vendor")]
    {
        use std::path::Path;

        fn copy_all(out_dir: &Path, assets: &[(&str, &str)]) -> std::io::Result<()> {
            let root = Path::new(env!("CARGO_MANIFEST_DIR")).join("../");
            for (_, path) in assets {
                let src = root.join(path);
                let dest = out_dir.join(path);
                if let Some(parent) = dest.parent() {
                    std::fs::create_dir_all(parent)?;
                }
                std::fs::copy(&src, &dest)?;
                println!("cargo:rerun-if-changed={}", src.display());
            }
            Ok(())
        }

        fn generate_rust_module(out_dir: &Path, assets: &[(&str, &str)]) -> std::io::Result<()> {
            let mut module = String::from("//! Auto-generated asset constants\\n\\n");
            for (name, path) in assets {
                module.push_str(&format!("pub const {}: &[u8] = include_bytes!(\\\"{}\\\");\\n", name, path));
            }
            std::fs::write(out_dir.join("rlvgl_assets.rs"), module)?;
            Ok(())
        }

        const ASSETS: &[(&str, &str)] = &[
        {% for a in assets %}
            ("{{ a.name }}", "{{ a.path }}"),
        {% endfor %}
        ];

        let out = std::path::PathBuf::from(std::env::var("OUT_DIR").unwrap());
        copy_all(&out, ASSETS)?;
        generate_rust_module(&out, ASSETS)?;
    }
    Ok(())
}
"#;
    tera.add_raw_template("Cargo.toml", CARGO_TOML)?;
    tera.add_raw_template("lib.rs", LIB_RS)?;
    tera.add_raw_template("README.md", README_MD)?;
    tera.add_raw_template("build.rs", BUILD_RS)?;

    let mut ctx = Context::new();
    ctx.insert("name", &crate_name);
    ctx.insert("groups", &groups);
    ctx.insert("assets", &embed_assets);

    fs::write(path.join("Cargo.toml"), tera.render("Cargo.toml", &ctx)?)?;
    fs::write(path.join("src/lib.rs"), tera.render("lib.rs", &ctx)?)?;
    fs::write(path.join("README.md"), tera.render("README.md", &ctx)?)?;
    fs::write(path.join("build.rs"), tera.render("build.rs", &ctx)?)?;

    println!("Scaffolded crate `{}`", crate_name);
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;
    use tempfile::tempdir;

    #[test]
    fn scaffold_generates_expected_files() {
        let dir = tempdir().unwrap();
        let crate_path = dir.path().join("test_assets");
        // Use a missing manifest path so defaults are applied.
        run(&crate_path, Path::new("missing.yml")).unwrap();

        let read = |p: &str| std::fs::read_to_string(crate_path.join(p)).unwrap();

        insta::assert_snapshot!("cargo_toml", read("Cargo.toml"));
        insta::assert_snapshot!("lib_rs", read("src/lib.rs"));
        insta::assert_snapshot!("readme_md", read("README.md"));
        insta::assert_snapshot!("build_rs", read("build.rs"));

        // Generate an empty `rlvgl_assets.rs` module and snapshot its contents.
        let out_dir = crate_path.join("out");
        std::fs::create_dir_all(&out_dir).unwrap();
        generate_rust_module(&out_dir, &[]).unwrap();
        let read_out = |p: &str| std::fs::read_to_string(out_dir.join(p)).unwrap();
        insta::assert_snapshot!("rlvgl_assets_rs", read_out("rlvgl_assets.rs"));
    }

    #[test]
    fn scaffold_cargo_publish_dry_run_succeeds() {
        let dir = tempdir().unwrap();
        let crate_path = dir.path().join("publishable_assets");
        run(&crate_path, Path::new("missing.yml")).unwrap();

        let status = std::process::Command::new("cargo")
            .arg("publish")
            .arg("--dry-run")
            .arg("--allow-dirty")
            .current_dir(&crate_path)
            .status()
            .expect("failed to run cargo publish");

        assert!(status.success());
    }

    #[test]
    fn scaffold_cargo_check_embed_and_vendor_succeed() {
        let dir = tempdir().unwrap();
        let crate_path = dir.path().join("buildable_assets");
        run(&crate_path, Path::new("missing.yml")).unwrap();

        let check = |features: &str| {
            let status = std::process::Command::new("cargo")
                .arg("check")
                .arg("--features")
                .arg(features)
                .current_dir(&crate_path)
                .status()
                .expect("failed to run cargo check");
            assert!(status.success());
        };

        check("embed");
        check("vendor");
    }

    #[test]
    fn embed_and_vendor_yield_identical_bytes() {
        use crate::{raw, scan, vendor};
        use image::{DynamicImage, Rgba, RgbaImage};
        use serde_yaml;

        let dir = tempdir().unwrap();
        let root = dir.path();

        // Create a tiny raw asset under the required root.
        let icons = root.join("icons");
        std::fs::create_dir_all(&icons).unwrap();
        let img = DynamicImage::ImageRgba8(RgbaImage::from_pixel(1, 1, Rgba([1, 2, 3, 4])));
        raw::encode_image(img, icons.join("test.raw")).unwrap();

        // Scan the asset and mark it with a license so vendoring succeeds.
        let manifest_path = root.join("manifest.yml");
        scan::run(root, &manifest_path).unwrap();
        let mut manifest: crate::manifest::Manifest =
            serde_yaml::from_str(&std::fs::read_to_string(&manifest_path).unwrap()).unwrap();
        manifest.assets[0].license = Some("MIT".to_string());
        std::fs::write(&manifest_path, serde_yaml::to_string(&manifest).unwrap()).unwrap();

        // Scaffold an assets crate and vendor the asset.
        let crate_path = root.join("test_assets");
        run(&crate_path, &manifest_path).unwrap();
        let out_dir = root.join("out");
        vendor::run(root, &manifest_path, &out_dir, &["MIT".into()], &[]).unwrap();

        // Embedded bytes should match vendored bytes for the same asset.
        let embed = std::fs::read(root.join("icons/test.raw")).unwrap();
        let vendored = std::fs::read(out_dir.join("icons/test.raw")).unwrap();
        assert_eq!(embed, vendored);
    }

    fn generate_rust_module(out_dir: &Path, assets: &[(&str, &str)]) -> std::io::Result<()> {
        let mut module = String::from("//! Auto-generated asset constants\n\n");
        for (name, path) in assets {
            module.push_str(&format!(
                "pub const {}: &[u8] = include_bytes!(\"{}\");\n",
                name, path
            ));
        }
        std::fs::write(out_dir.join("rlvgl_assets.rs"), module)
    }
}
