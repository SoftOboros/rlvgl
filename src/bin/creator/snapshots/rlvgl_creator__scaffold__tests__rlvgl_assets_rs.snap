---
source: src/bin/creator/scaffold.rs
expression: "read_out(\"rlvgl_assets.rs\")"
---
//! Auto-generated asset constants

